test/test_sparse.py

```diff
@@ -8,78 +8,85 @@
 from common_nn import TEST_CUDA
 from numbers import Number
 
-# triplet := (index type, value type, sparse type)
-cpu_triplet = (
-    torch.LongTensor,
-    torch.DoubleTensor,
-    torch.sparse.DoubleTensor)
 
-if TEST_CUDA:
-    cuda_triplet = (
-        torch.cuda.LongTensor,
-        torch.cuda.DoubleTensor,
-        torch.cuda.sparse.DoubleTensor)
+def cpu_only(inner):
+    def outer(self, *args, **kwargs):
+        unittest.skipIf(self.is_cuda, "Test is CPU-only")(inner)(self, *args, **kwargs)
+    return outer
 
 
 class TestSparse(TestCase):
 
-    @staticmethod
-    def _gen_sparse(d, nnz, with_size, is_cuda=False):
+    def setUp(self):
+        # These parameters control the various ways we can run the test.
+        # We will subclass and override this method to implement CUDA
+        # tests
+        self.is_cuda = False
+        self.IndexTensor = torch.LongTensor
+        self.ValueTensor = torch.DoubleTensor
+        self.SparseTensor = torch.sparse.DoubleTensor
+
+    def _gen_sparse(self, d, nnz, with_size):
+        # TODO: Consider implementing this in the CUDA case by directly
+        # performing the operations on the GPU.  You won't be able to
+        # use torch.rand/torch.randn in this case because they are
+        # CPU-only.  If you do this, you can remove the is_cuda branch
+        # at the end.
+
         if isinstance(with_size, Number):
             v = torch.randn(nnz)
             i = (torch.rand(d, nnz) * with_size).type(torch.LongTensor)
             x = torch.sparse.DoubleTensor(i, v)
         else:
+            # Generate a sparse tensor with d sparse dimensions; the
+            # rest the dimensions with_size[d:] are dense.
             v_size = [nnz] + list(with_size[d:])
             v = torch.randn(*v_size)
             i = torch.rand(d, nnz) * \
                 torch.Tensor(with_size[:d]).repeat(nnz, 1).transpose(0, 1)
             i = i.type(torch.LongTensor)
             x = torch.sparse.DoubleTensor(i, v, torch.Size(with_size))
 
-        if is_cuda:
+        if self.is_cuda:
             return x.cuda(), i.cuda(), v.cuda()
         else:
             return x, i.clone(), v.clone()
 
-    def _test_basic(self, is_cuda):
-        x, i, v = self._gen_sparse(3, 10, 100, is_cuda)
+    def randn(self, *args, **kwargs):
+        # TODO: Maybe do this directly on GPU
```

Comment:
Yes. But then I thought, "Maybe they did it this way on purpose, so that the random seed behavior was consistent over CPU and GPU." But I'm probably overthinking it :)